src/API.jl

```diff
@@ -224,32 +225,28 @@ function free(ctx::Context, pkgs::Vector{PackageSpec}; kwargs...)
     return
 end
 
-test(;kwargs...)                                  = test(PackageSpec[]; kwargs...)
-test(pkg::Union{AbstractString, PackageSpec}; kwargs...)  = test([pkg]; kwargs...)
-test(pkgs::Vector{<:AbstractString}; kwargs...)           = test([PackageSpec(pkg) for pkg in pkgs]; kwargs...)
-test(pkgs::Vector{PackageSpec}; kwargs...)        = test(Context(), pkgs; kwargs...)
-
-function test(ctx::Context, pkgs::Vector{PackageSpec}; coverage=false, kwargs...)
-    pkgs = deepcopy(pkgs)  # deepcopy for avoid mutating PackageSpec members
+test(;kwargs...)                                         = test(PackageSpec[]; kwargs...)
+test(pkg::Union{AbstractString, PackageSpec}; kwargs...) = test([pkg]; kwargs...)
+test(pkgs::Vector{<:AbstractString}; kwargs...)          = test([PackageSpec(pkg) for pkg in pkgs]; kwargs...)
+test(pkgs::Vector{PackageSpec}; kwargs...)               = test(Context(), pkgs; kwargs...)
+function test(ctx::Context, pkgs::Vector{PackageSpec};
+              coverage=false, test_fn=nothing, kwargs...)
```

Comment:
Why is `test_fn` needed?

Reply:
I used it for testing purposes. I allows one to hook into the `test` environment. We can also put it in `Context` or have some other way of testing within the `test` environment. I just needed something to be in place: I'm not super convinced of a superior alternative.